src/geometry/euclidean.lean

```diff
@@ -683,4 +683,230 @@ begin
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p1 : P} (p2 : P)
+    (hp1 : p1 ∈ s) :
+  p1 -ᵥ orthogonal_projection hn hc p2 ∈ s.direction :=
+vsub_mem_direction hp1 (orthogonal_projection_mem hn hc p2)
+
+/-- A point equals its orthogonal projection if and only if it lies in
+the subspace. -/
+lemma orthogonal_projection_eq_self_iff {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  orthogonal_projection hn hc p = p ↔ p ∈ s :=
+begin
+  split,
+  { exact λ h, h ▸ orthogonal_projection_mem hn hc p },
+  { intro h,
+    have hp : p ∈ ((s : set P) ∩ mk' p s.direction.orthogonal) := ⟨h, self_mem_mk' p _⟩,
+    rw [inter_eq_singleton_orthogonal_projection hn hc p, set.mem_singleton_iff] at hp,
+    exact hp.symm }
+end
+
+/-- The distance between a point and its orthogonal projection is
+nonzero if it does not lie in the subspace. -/
+lemma dist_orthogonal_projection_ne_zero_of_not_mem {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p : P} (hp : p ∉ s) :
+  dist p (orthogonal_projection hn hc p) ≠ 0 :=
+begin
+  intro hz,
+  rw dist_eq_zero at hz,
+  symmetry' at hz,
+  rw orthogonal_projection_eq_self_iff hn hc _ at hz,
+  exact hp hz
+end
+
+/-- Subtracting `p` from its `orthogonal_projection` produces a result
+in the orthogonal direction. -/
+lemma orthogonal_projection_vsub_mem_direction_orthogonal {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  orthogonal_projection hn hc p -ᵥ p ∈ s.direction.orthogonal :=
+orthogonal_projection_fn_vsub_mem_direction_orthogonal hn hc p
+
+/-- Subtracting the `orthogonal_projection` from `p` produces a result
+in the orthogonal direction. -/
+lemma vsub_orthogonal_projection_mem_direction_orthogonal {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) (p : P) :
+  p -ᵥ orthogonal_projection hn hc p ∈ s.direction.orthogonal :=
+direction_mk' p s.direction.orthogonal ▸
+  vsub_mem_direction (self_mem_mk' _ _) (orthogonal_projection_mem_orthogonal hn hc p)
+
+/-- Adding a vector to a point in the given subspace, then taking the
+orthogonal projection, produces the original point if the vector was
+in the orthogonal direction. -/
+lemma orthogonal_projection_vadd_eq_self {s : affine_subspace ℝ V P}
+    (hn : (s : set P).nonempty) (hc : is_complete (s.direction : set V)) {p : P} (hp : p ∈ s)
+    {v : V} (hv : v ∈ s.direction.orthogonal) : orthogonal_projection hn hc (v +ᵥ p) = p :=
+begin
+  have h := vsub_orthogonal_projection_mem_direction_orthogonal hn hc (v +ᵥ p),
+  rw [vadd_vsub_assoc, submodule.add_mem_iff_right _ hv] at h,
+  exact (eq_of_vsub_eq_zero V
+          (submodule.disjoint_def.1
+            s.direction.orthogonal_disjoint
+            _
+            (vsub_mem_direction hp (orthogonal_projection_mem hn hc (v +ᵥ p)))
+            h)).symm
```

Comment:
I found the following more pleasant to parse, but I'm not convinced it's the best:
```suggestion
  refine (eq_of_vsub_eq_zero V _).symm,
  refine submodule.disjoint_def.1 s.direction.orthogonal_disjoint _ _ h,
  exact vsub_mem_direction hp (orthogonal_projection_mem hn hc (v +ᵥ p))
```

Reply:
I probably had something like that originally before golfing a sequence of tactics at the end of the proof into a term. I'm not sure if there's any convention about the extent to which to convert parts of proofs into terms like that?